internal/catalog/internal/controllers/endpoints/controller.go

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package endpoints
+
+import (
+	"context"
+	"sort"
+
+	"github.com/hashicorp/consul/internal/catalog/internal/controllers/workloadhealth"
+	"github.com/hashicorp/consul/internal/catalog/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	endpointsMetaManagedBy = "managed-by-controller"
+)
+
+// The WorkloadMapper interface is used to provide an implementation around being able
+// to map a watch even for a Workload resource and translate it to reconciliation requests
+type WorkloadMapper interface {
+	// MapWorkload conforms to the controller.DependencyMapper signature. Given a Workload
+	// resource it should report the resource IDs that have selected the workload.
+	MapWorkload(context.Context, controller.Runtime, *pbresource.Resource) ([]controller.Request, error)
+
+	// TrackIDForSelector should be used to associate the specified WorkloadSelector with
+	// the given resource ID. Future calls to MapWorkload
+	TrackIDForSelector(*pbresource.ID, *pbcatalog.WorkloadSelector)
+
+	// UntrackID should be used to inform the tracker to forget about the specified ID
+	UntrackID(*pbresource.ID)
+}
+
+// ServiceEndpointsController creates a controller to perform automatic endpoint management for
+// services.
+func ServiceEndpointsController(workloadMap WorkloadMapper) controller.Controller {
+	if workloadMap == nil {
+		panic("No WorkloadMapper was provided to the ServiceEndpointsController constructor")
+	}
+
+	return controller.ForType(types.ServiceEndpointsType).
+		WithWatch(types.ServiceType, controller.ReplaceType(types.ServiceEndpointsType)).
+		WithWatch(types.WorkloadType, workloadMap.MapWorkload).
+		WithReconciler(newServiceEndpointsReconciler(workloadMap))
+}
+
+type serviceEndpointsReconciler struct {
+	workloadMap WorkloadMapper
+}
+
+func newServiceEndpointsReconciler(workloadMap WorkloadMapper) *serviceEndpointsReconciler {
+	return &serviceEndpointsReconciler{
+		workloadMap: workloadMap,
+	}
+}
+
+// Reconcile will reconcile one ServiceEndpoints resource in response to some event.
+func (r *serviceEndpointsReconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
```

Comment:
Should we add some logging to this controller as well?

Reply:
Yep, I can do that.